win-asio: Add ASIO host for obs-studio

pkviet opened the pull request:
### Description
ASIO is a driver standard developed by Steinberg for low-latency audio. It is commonly used in pro audio.
It was unfortunately not possible for a long time to write a native ASIO host in obs because of license issues.
This has changed thanks to a partnership between the project and Yamaha Steinberg:
https://www.steinberg.net/press/2025/obs-collaboration/
The ASIO sdk is now released under dual licensing, including GPL v3 which is compatible with obs-studio.

**Historical aside.** When I got involved with obs coding, after working on surround sound support, my first plugin
endeavor was to write an asio source, which was developed eventually in collaboration with Andersama. It was a great learning experience. The ASIO sdk is notoriously tricky with lots of pitfalls for hosts and drivers alike.  It was so daunting that at that time (early 2018) we relied on third-party libraries (RTAudio, Bassasio, portaudio) and now JUCE. The last incarnation of our asio plugin for obs (based on JUCE) was mostly coded by Andersama. It has been superseded in many ways by another third-party plugin based also on JUCE, atkaudio which has lot of capabilities but again relies on a third-party library. And anyway our ultimate goal had always been to add support for ASIO into obs.

**This PR brings ASIO support in obs-studio natively: no third-party library is used but Steinberg SDK is leveraged directly.** 
For the host implementation, we have taken cues from JUCE ASIO host which documents a lot of pitfalls.

Our ASIO host provides:
- an asio source where inputs from asio devices can be captured with also channel selection.
- an asio output.

The asio output in turn allows:
- track monitoring with channel selection too,
- a special monitoring track which automatically mixes all monitored sources and can be selected in the asio output.

#### Technical note
The asio callback processes both audio from host to device and from device to host at the same time. The main difference with wasapi, is that any monitored audio sent to the device is not automatically mixed; it is the responsibility of the host to do the mixing. That is why asio can not be seamlessly integrated into libobs/audio-monitoring and that we resorted to creating an obs-output for asio.
This output is controlled through a frontend plugin which instead of adding an entry into the tool menu (as is customary) is callable from Settings > Audio alongside monitoring devices (wasapi on windows).

#### Screenshots:
**ASIO source**
<img width="630" height="667" alt="obs64_2025-09-02_17-25-32" src="https://github.com/user-attachments/assets/6982d0d1-d092-420b-92f4-a694171a4c05" />

**ASIO source with channel selection**
<img width="630" height="667" alt="obs64_2025-09-02_17-25-51" src="https://github.com/user-attachments/assets/b8c20f3e-b62d-4067-880c-4fdfcd7aaed7" />

**ASIO source with device selection**
<img width="630" height="667" alt="obs64_2025-09-02_17-25-58" src="https://github.com/user-attachments/assets/da6ccb84-35fb-42d9-9145-7afa138a3c6e" />

**ASIO monitoring integrated to audio settings**
<img width="919" height="685" alt="obs64_2025-10-15_20-46-35" src="https://github.com/user-attachments/assets/75d52f8b-f99a-4b8d-a235-5e26729d86a1" />

**ASIO monitoring panel**
<img width="983" height="681" alt="obs64_2025-10-15_20-52-18" src="https://github.com/user-attachments/assets/dc89cac7-fe37-4f6c-9592-c3461b2698ea" />

**ASIO monitoring panel with track channel selection. Not the special monitoring track**
<img width="977" height="936" alt="obs64_2025-10-15_20-52-40" src="https://github.com/user-attachments/assets/460cfde5-1e92-403c-94f8-8565abbbe523" />

### Motivation and Context
ASIO brings the following advantages:
- low-latency capture,
- low-latency monitoring (the wasapi monitoring has long been a source of user complaints),
- capability to do complex channel routing.

### How Has This Been Tested?
Tested on windows 11 pro with several asio devices:
- Steinberg IXO 22
- Steinberg UR22C
- ESI U22XT
- Behringer XR18
- Tascam US-122 MKII
- virtual drivers such as Rearoute from Reaper, Dante virtual card

**Testers are welcome to report their results with their interface and I'll update the list.**

### Types of changes
- New feature (non-breaking change which adds functionality) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

pkviet committed:
win-asio: Add ASIO source & output

This adds:
- an ASIO Input source, allowing audio capture of devices using
Steinberg ASIO audio SDK on Windows. This allows low latency capture.
- an ASIO output, which will allow routing audio from OBS to any ASIO
device. A per channel routing is provided.
The UI to setup the ASIO output is provided in a later commit.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
libobs: Add ASIO monitoring track

On windows, this adds an additional audio track (Track 7) which is used
to output a mix of monitored sources. Currently the mixing is left to
WASAPI on Windows for WASAPI monitoring devices.
But ASIO SDK doesn't allow multiple clients (although individual drivers
 might allow it) and therefore can not mix them. So we have to do the
mixing on OBS side.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
win-asio: Enable monitoring track for output

This enables a monitoring track for output.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
UI: Add ASIO output frontend plugin

This adds an ASIO output entry in Tools Menu allowing to setup audio
output from obs to an ASIO device.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
UI: Add to Settings > Audio ASIO monitoring

This adds a QToolButton to the Audio Settings on Windows.
This triggers in turn a dedicated settings for the ASIO monitoring
output.
If no ASIO driver is detected in the system, the panel displays an
explanatory message.
The panel allows to:
- select an ASIO monitoring device;
- for each output channel of the ASIO device, one can select any channel
from any of the 6 tracks or from the monitoring mix.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet commented:
**Update**
- Fixed a lot of bugs; thanks to @prgmitchell , indrit (on discord), @WizardCM 
- On @WizardCM suggestion, i added a message when a driver fails to load (usually because the device is disconnected,
or just because of flaky drivers - there's a lot of them in ASIO landscape , or because the ASIO driver is being used by another host.)
<img width="829" height="375" alt="obs64_2025-10-17_21-01-58" src="https://github.com/user-attachments/assets/dd374257-d00b-4374-bc5f-4072b834a36d" />

pkviet commented:
**Update 10-19-25*
Fixed some bugs with output.
Factored a lot of the input and output functions.
Fixed a bug with drivers which set a different default sample rate than that from obs

PatTheMav commented on frontend/plugins/CMakeLists.txt:
Probably makes sense to update this whole file to use the `add_obs_plugin` CMake function instead and use `add_obs_plugin(asio-output-ui PLATFORMS WINDOWS)` in this case specifically.

pkviet replied:
done (but curtailed to asio to avoid having changes to unrelated plugins bundled in this PR)

PatTheMav commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1260 @@
+	return requested_size;
+}
+
+//=============================================================
```
If the file is large/complex enough to require comments like this to "segment off" parts of it, then the file should be split up into separate source files instead.

PatTheMav commented on plugins/CMakeLists.txt:
```diff
@@ -57,7 +57,6 @@ add_obs_plugin(mac-virtualcam PLATFORMS MACOS)
 add_obs_plugin(nv-filters PLATFORMS WINDOWS)
 
 check_obs_browser()
-
```
Unrelated remove of newline.

PatTheMav reviewed (commented):
I feel like I need to see the code "in action" to better gauge its correctness. I there something like an "ASIO virtual device" or something I can install in my VM and then use this plugin with it?

pkviet commented on plugins/win-asio/asio.h:
This is a tiny part of the asio sdk, which is available as a zip and not pullable from github. The full sdk is much larger but i kept only what's required to get the COM interface working.
So dunno if it's worth getting that in obs-deps

pkviet commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,201 @@
+	bool is_open;
+	bool is_started;
+	bool driver_failure;
+	volatile bool need_to_reset;
```
Ok i wasn't aware that the volatile qualifier was not compulsory and would be automatically upgraded.
Currently the bool is used only in our atomics functions in ui thread and in the driver thread, so I assume it is then safe to drop the qualifier.
But the question is then whether it leads to adverse consequences to keep it?
Your observation would then extend to other places to the code base.
I merely followed the current usage without questioning it due to lack of litteracy that i will readily confess, but it seems i am not alone in that respect.
Is the question you ask of great import or just nitpicking ? Honest question.

pkviet commented:
### General architecture
<img width="1254" height="834" alt="VISIO_2025-11-09_11-09-43" src="https://github.com/user-attachments/assets/2aa5732b-fec7-4e58-ad75-5f5596ce428d" />

All the audio processing is done in the own ASIO driver thread, NOT obs audio thread. This ensures the lowest latency. 
One big difference with WASAPI is that both input and output buffers are processed concurrently.
On OBS side we need to ensure there is sufficient data to fill the buffers to avoid xruns. This is done through 2 deques, one on the capture side (ASIO Input source) and another on the play side (ASIO Output). 
The **play deque** is filled in the audio thread by OBS with output audio from tracks or from monitoring.
The **capture deque** is filled in the asio driver thread but delivered to obs on the audio thread at each audio tick (1024 samples usually).
As already explained above, the ASIO driver does no mixing, so it is not possible to push separately all monitored sources; they must be mixed on our side. This is why a special "monitoring" track has been added. When compared with WASAPI, we have as a bonus that the monitored sources have the same sync as in outputs since they are processed in the obs main audio callback at the same time.

#### The main ASIO SDK issue:

- this old sdk (it dates back to 1996, so pre-historic times) assumes that the host is a DAW and that it will load a single driver at a time.
- conversely, the drivers are also supposed to be single client to minimize latency and not split the processing among several hosts. (A few ASIO drivers fortunately are now multi-client, that is, allow concurrent use by several hosts at the same time: ex: RME or Lynx.)
- As a result of these limitations, it will not in general be possible to open an ASIO interface both in obs and a DAW. Workarounds do exist : for instance use Rearoute virtual ASIO driver, provided by Cockos Reaper.
- The single driver host architecture of the SDK is a big issue for OBS because we need a multi-client host; a setup with a single ASIO device + one virtual ASIO driver such as VB-cable or ReaRoute allowing routing from DAWS, is a reasonable scenario that we need to support.
- We had therefore to work around the ASIO callbacks because they don't pass a 'this' pointer allowing us to identify the ASIO driver. Our solution was to template callbacks, each driver having its own set.

#### The scanning phase.
The scan just checks the registry to list all installed ASIO drivers.
There is no check at that time of whether a device is actually connected or not.
The SDK has no way for us to check the connection state apart from actually loading the device.
There is then an error message but unfortunately the SDK has no explicit message regarding the disconnected state ...
If a device is erroring during loading, we disable its audio processing. But we don't discard it from the devices list because it could be: (1) the device is being concurrently used by another host, like a DAW; (2) a genuine disconnection, the USB cable being loose ... So  obs should stick to the source settings, to allow the user to recover their setup.

#### The loading phase.
The driver is loaded during the creation phase when a driver is picked by the source.
During this loading some basic tests are done (in asio_device_test function) to ensure the device will stream properly during the real operating.
The driver is then opened (asio_device_open), meaning that all settings are applied, correct sample rate, buffer size, buffers are created... The device is started once all this initialization is done, meaning that processing can start. If audio callbacks are not triggered by the driver within 300 ms, the driver is stopped.

#### The streaming phase.
The driver calls repeatedly the bufferSwitch callback; this is where the main processing logic is done on our side (asio_device_process_buffer): there we copy in the 'capture deque' the buffers coming from the device, and we copy the 'play deque' into the output buffers of the device.

pkviet committed:
libobs: Don't silence the extra monitoring track for ASIO

On Windows, we provide an extra track which mixes all monitored sources.
If a monitored source is muted, it is removed from the usual monitoring
callbacks (wasapi-output.c ...) which can make sense given the possible
interferences with Desktop Audio.
But for ASIO drivers, there is no such issue since there is no play
capture. We therefore enable a muted source to be monitored.
Since it is muted, it won't appear in the mix delivered to streams or
recordings but it will still be heard on the monitoring ASIO device.
This is quite in line with a usage where one prepares some audio before
putting it online. Currently with WASAPI, this requires either a
hardware mixer or a sound card with several outputs, or several devices.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
plugins: Enable extra ASIO monitoring track for transitions and slideshow

Modifies obs-transitions and image-source.
This allows the audio of transitions such as stingers and slideshows to
be heard on the ASIO monitoring track.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
libobs: Audio routing of MONITOR_ONLY sources

Currently, sources which have the monitoring_type MONITOR_ONLY are not
pushed into audio_input_buf so they are not available to outputs.
So this makes them unavailable to ASIO monitoring.
To fix that, we systematically push them to audio_input_buf but silence
the source for all tracks except the extra ASIO monitoring track in the
apply_volume call, further down in the audio pipeline.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
obs-transitions: Set tracks for stinger

Currently the stinger is output on all tracks.
When MONITOR_ONLY is picked, we explicitly set the mixers bitwise to 0
so that the audio never goes to outputs except the extra ASIO track.

Signed-off-by: pkv <pkv@obsproject.com>

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+	return nullptr;
+}
+
+#define MAX_DEVICE_CHANNELS 32
```
Consider moving this up with the other globals for this file and putting them into an anonymous namespace.
```suggestion
constexpr int MAX_DEVICE_CHANNELS = 32;
```

pkviet replied:
done, except for the namespace; i instead renamed into MAX_ASIO_DEVICE_CHANNELS to avoid collisions

RytoEX replied:
If it can be placed in an anonymous namespace (i.e., it is not used outside of this file), it should be. The benefit isn't just about namespace collision, but internal/external linkage. An anonymous namespace is akin to using the `static` keyword, but is considered the modern C++ preferred method for this.

pkviet replied:
ok, i've put it into an anonymous namespace

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+		if (!obsSettingsDialog)
+			settingsDialog_ = new ASIOSettingsDialog(mainWindow, context.output, context.settings);
+		else
+			settingsDialog_ = new ASIOSettingsDialog(obsSettingsDialog, context.output, context.settings);
```
Prefer:
```suggestion
		if (!obsSettingsDialog) {
			settingsDialog_ = new ASIOSettingsDialog(mainWindow, context.output, context.settings);
		} else {
			settingsDialog_ = new ASIOSettingsDialog(obsSettingsDialog, context.output, context.settings);
		}
```

pkviet replied:
done

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+	if (obsSettingsDialog) {
+		QRect settingsRect = obsSettingsDialog->geometry();
+		QRect asioRect = settingsDialog_->geometry();
+		QPoint newPos(settingsRect.right() + 100, settingsRect.top());
```
Why `+ 100`?

pkviet replied:
that's to avoid obscuring the Settings window since the asio output window is integrated into the Settings through a call in a button box.

RytoEX replied:
Then we should not use a magic number but a named constant/constexpr to make it clear what this is and why we need to add 100 to the position value.

pkviet replied:
ok added SHIFT_WINDOW = 100 as a constexpr in the anonymous namespace

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+		QRect desktopRect = screen->availableGeometry();
+		if (newPos.x() + asioRect.width() > desktopRect.right())
+			newPos.setX(desktopRect.right() - asioRect.width());
+		settingsDialog_->move(newPos);
```
```suggestion
		if (newPos.x() + asioRect.width() > desktopRect.right()) {
			newPos.setX(desktopRect.right() - asioRect.width());
		}
		settingsDialog_->move(newPos);
```

pkviet replied:
done

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+	} else {
+		blog(LOG_INFO, "Failed to create ASIO output");
+		// We add the UI even if there is no output, to display a text saying ASIO is disabled
+		// so that users know that ASIO exists ...
```
```suggestion
		// We add the UI even if there is no output to display a text saying ASIO is disabled
		// so that users know that ASIO exists.
```

pkviet replied:
done

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#include "ASIOSettingsDialog.h"
+#include <obs-frontend-api.h>
+#include <obs-module.h>
+#include <util/util.hpp>
+#include <util/platform.h>
+#include <QMainWindow>
+#include <QScreen>
```
Break after the license header and before the includes. Separate incluedes.

```suggestion
******************************************************************************/

#include "ASIOSettingsDialog.h"

#include <obs-frontend-api.h>
#include <obs-module.h>
#include <util/platform.h>
#include <util/util.hpp>

#include <QMainWindow>
#include <QScreen>
```

pkviet replied:
done

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+/******************************************************************************
+    Copyright (C) 2022-2025 pkv <pkv@obsproject.com>
+
+    This file is part of win-asio.
```
This file is actually part of asio-output-ui or obs-studio.

pkviet replied:
changed into: This file is part of asio-output-ui which requires win-asio.

RytoEX commented on frontend/plugins/asio-output-ui/ASIOSettingsDialog.h:
```diff
@@ -0,0 +1,44 @@
+#include <properties-view.hpp>
+#include <memory>
+#include <QDialog>
+#include "./forms/ui_output.h"
```
Not sure if the path specificity is required?

pkviet replied:
indeed; done

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,199 @@
+struct asio_ui_output context = {0};
+bool output_running = false;
+ASIOSettingsDialog *settingsDialog_ = nullptr;
+std::string g_currentDeviceName;
```
Perhaps an anonymous namespace?
```suggestion
// We use a global context for asio output because we allow a single output device.
namespace {
struct asio_ui_output context = {0};
bool output_running = false;
ASIOSettingsDialog *settingsDialog_ = nullptr;
std::string g_currentDeviceName;
} // namespace
```

pkviet replied:
output_running and g_currentDeviceName would not be available as extern in ASIOSettingsDialog.cpp; the latter does not compile with your suggestion. The odds of collisions with these symbols should be 0, if that is the concern ?

RytoEX replied:
Using an anonymous namespace is not entirely about namespace collisions, though that is a benefit. Anonymous namespaces may also reduce linking time. In any case, it is considered a C++ best practice to avoid declaring identifiers in the global namespace. See:
* https://google.github.io/styleguide/cppguide.html#Internal_Linkage
* https://chromium.googlesource.com/chromium/src/+/0ef96750ee5b58aa95e43b25e166ab1b0b747711/styleguide/c++/c++.md#unnamed-namespaces
* https://github.com/cpp-best-practices/cppbestpractices/blob/d57b14ec6d7f13f0af65f9bde62aa3ebd5fb588a/03-Style.md#always-use-namespaces

If some are still required via extern, you can leave those in the global namespace for now, though ideally we wouldn't have to have them there either.

pkviet replied:
ok i've kept in the global namespace only those used elsewhere and moved the others to the anonymous namespace

RytoEX commented on libobs/obs-source.c:
```diff
@@ -1647,12 +1648,10 @@ static void source_output_audio_data(obs_source_t *source, const struct audio_da
+	if (push_back && source->audio_ts)
+		source_output_audio_push_back(source, &in);
+	else
+		source_output_audio_place(source, &in);
```
```suggestion
	if (push_back && source->audio_ts) {
		source_output_audio_push_back(source, &in);
	} else {
		source_output_audio_place(source, &in);
	}
```

pkviet replied:
done

RytoEX commented on libobs/obs-source.c:
```diff
@@ -5587,10 +5595,30 @@ void obs_source_set_monitoring_type(obs_source_t *source, enum obs_monitoring_ty
+
+#ifdef _WIN32
+	// On windows, assign to the extra asio monitoring track (track 7) all sources which have not type
+	// OBS_MONITORING_TYPE_NONE.
```
```suggestion
	// On Windows, assign to the extra ASIO monitoring track (track 7) all sources which have not type
	// OBS_MONITORING_TYPE_NONE.
```

pkviet replied:
done

RytoEX commented on libobs/obs-source.c:
```diff
@@ -5587,10 +5595,30 @@ void obs_source_set_monitoring_type(obs_source_t *source, enum obs_monitoring_ty
 enum obs_monitoring_type obs_source_get_monitoring_type(const obs_source_t *source)
 {
+#ifdef _WIN32
+	// If type is OBS_MONITORING_TYPE_NONE, unselect the extra asio monitoring track (track 7) on windows.
```
```suggestion
	// If type is OBS_MONITORING_TYPE_NONE, unselect the extra ASIO monitoring track (track 7) on Windows.
```

pkviet replied:
done

RytoEX commented on plugins/obs-transitions/transition-stinger.c:
```diff
@@ -137,6 +137,12 @@ static void stinger_update(void *data, obs_data_t *settings)
+	if (s->monitoring_type == OBS_MONITORING_TYPE_MONITOR_ONLY)
+		obs_source_set_audio_mixers(s->media_source, 0);
+	else
+		obs_source_set_audio_mixers(s->media_source, 0xFF);
```
```suggestion
	// If the stinger is set to monitor only, don't send audio to output mixers.
	if (s->monitoring_type == OBS_MONITORING_TYPE_MONITOR_ONLY) {
		obs_source_set_audio_mixers(s->media_source, 0);
	} else {
		obs_source_set_audio_mixers(s->media_source, 0xFF);
	}
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/data/locale/en-US.ini:
```diff
@@ -0,0 +1,113 @@
+ASIO.Driver.Error="The driver failed to load. Consult the log for possible hints."
+ASIO.Driver.None="No ASIO audio driver installed."
+Control.Panel="ASIO Device Control Panel"
+Control.Panel.Hint="If you change any settings of the device, like the buffer, while it's streaming, hit the reload button to reload the driver in case of glitches (due to bad drivers not messagin changes to hosts)."
```
```suggestion
Control.Panel.Hint="If you change any settings of the device, like the buffer, while it's streaming, hit the reload button to reload the driver in case of glitches (due to bad drivers not messaging changes to hosts)."
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/data/locale/en-US.ini:
```diff
@@ -0,0 +1,113 @@
+Control.Panel="ASIO Device Control Panel"
+Control.Panel.Hint="If you change any settings of the device, like the buffer, while it's streaming, hit the reload button to reload the driver in case of glitches (due to bad drivers not messagin changes to hosts)."
+Reset.Device="Reload driver"
+Reset.Device.Hint="Reloads manually the driver. This is useful in case of driver settings being modified (sample rate, buffer, ...)"
```
```suggestion
Reset.Device.Hint="Reloads manually the driver. This is useful in case of driver settings being modified (sample rate, buffer, etc.)"
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-callbacks.c:
```diff
@@ -0,0 +1,83 @@
+/* ASIO callbacks are quite a pain because they don't store the 'this' pointer of the COM driver. In c++ we could do
+ * some template tricks as done by JUCE. But in C, we're more restricted, hence the use of these macros.
+ * The number of callback sets is the number of supported devices. So up to 16, which is plenty enough !
+ */
```
```suggestion
/* ASIO callbacks are quite a pain because they don't store the 'this' pointer of the COM driver. In C++, we could do
 * some template tricks as done by JUCE. But in C, we're more restricted, hence the use of these macros.
 * The number of callback sets is the number of supported devices. So up to 16, which is plenty enough!
 */
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-common.h:
```diff
@@ -0,0 +1,45 @@
+#define ASIO_LOG(level, format, ...)  \
+	blog(level, "[asio_device '%s']: " format, (dev && dev->device_name) ? dev->device_name : "(null)", ##__VA_ARGS__)
+
+#define ASIO_LOG2(level, format, ...) \
```
Perhaps a more informative name? Without context, I don't know why one would use `ASIO_LOG2`.

pkviet replied:
ASIO_LOG is for asio_device while ASIO_LOG2 is for asio_device_list which is the device scanner.
This is left unexplained because the code is self-explanatory ?
`blog(level, "[asio_device '%s']: "`

vs

`blog(level, "[asio_device_list]: "`

RytoEX replied:
My concern is being able to understand it quickly 5 years from now.

pkviet replied:
ok changed to:

- ASIO_DEVICE_LOG
- ASIO_SCANNER_LOG

RytoEX commented on plugins/win-asio/asio-common.h:
```diff
@@ -0,0 +1,45 @@
+#define info(format, ...)  ASIO_LOG(LOG_INFO, format, ##__VA_ARGS__)
+#define info2(format, ...) ASIO_LOG2(LOG_INFO, format, ##__VA_ARGS__)
+#define error(format, ...) ASIO_LOG(LOG_ERROR, format, ##__VA_ARGS__)
+#define error2(format, ...) ASIO_LOG2(LOG_ERROR, format, ##__VA_ARGS__)
```
Perhaps a more informative name than `info2` and `error2`?

pkviet replied:
changed to infoscanner and errorscanner

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#include "asio-device-list.h"
+#include <obs-module.h>
+#include <objbase.h>
+#include <string.h>
+#include <strsafe.h>
+#include <windows.h>
```
```suggestion
******************************************************************************/

#include "asio-device-list.h"

#include <obs-module.h>

#include <objbase.h>
#include <string.h>
#include <strsafe.h>
#include <windows.h>
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+#include <windows.h>
+
+// blacklist taken from JUCE to which I added the crappy Realtek ASIO, which gave me only headaches
+static const char *blacklisted_drivers[] = {"ASIO DirectX Full Duplex", "ASIO Multimedia Driver", "Realtek ASIO", NULL};
```
```suggestion
// Blacklist taken from JUCE to which I added Realtek ASIO.
static const char *blacklisted_drivers[] = {"ASIO DirectX Full Duplex", "ASIO Multimedia Driver", "Realtek ASIO", NULL};
```

pkviet replied:
done; but i liked the derogatory comment because that piece of crap did bsod my pc a couple of times.

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+}
+
+const char *asio_device_list_get_name(const struct asio_device_list *list, size_t index)
+{
+	if (!list || index >= list->count)
+		return NULL;
+	return list->drivers[index].name;
```
```suggestion
	if (!list || index >= list->count)
		return NULL;

	return list->drivers[index].name;
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+	return -1;
+}
+
+int asio_device_list_get_index_from_driver_name(const struct asio_device_list *list, const char *name)
+{
+	if (!list || !name)
+		return -1;
+	if (!list->has_scanned)
+		return -1;
+	for (size_t i = 0; i < list->count; ++i) {
```
```suggestion
	if (!list || !name)
		return -1;

	if (!list->has_scanned)
		return -1;

	for (size_t i = 0; i < list->count; ++i) {
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+	return -1;
+}
+
+struct asio_device *asio_device_list_attach_device(struct asio_device_list *list, const char *name)
+{
+	if (!list || !name)
+		return NULL;
+	if (!list->has_scanned)
+		return NULL;
+	int index = asio_device_list_get_index_from_driver_name(list, name);
```
```suggestion
	if (!list || !name)
		return NULL;

	if (!list->has_scanned)
		return NULL;

	const int index = asio_device_list_get_index_from_driver_name(list, name);
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,202 @@
+					return current_asio_devices[j];
+			}
+		}
+		int free_slot = find_free_driver_slot();
```
```suggestion
		const int free_slot = find_free_driver_slot();
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.h:
```diff
@@ -0,0 +1,42 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#pragma once
+#include "asio-device.h"
+#include <stdbool.h>
+#include <windows.h>
```
```suggestion
******************************************************************************/

#pragma once

#include "asio-device.h"

#include <stdbool.h>
#include <windows.h>
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,140 @@
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#pragma once
+#include "asio-common.h"
+#include "asio-format.h"
+#include "iasiodrv.h"
+#include "asio.h"
+#include <util/darray.h>
+#include <util/deque.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <windows.h>
```
```suggestion
******************************************************************************/

#pragma once

#include "asio-common.h"
#include "asio-format.h"
#include "asio.h"
#include "iasiodrv.h"

#include <util/darray.h>
#include <util/deque.h>

#include <stdbool.h>
#include <stdint.h>
#include <windows.h>
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,140 @@
+	 * created by the host. */
+	int slot_number;
+
+	/* Capture Audio (device ==> OBS).
```
```suggestion
	/* Capture Audio (device ==> OBS)
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,140 @@
+	int slot_number;
+
+	/* Capture Audio (device ==> OBS).
+	 * Each device will stream audio to a number of obs asio sources acting as audio clients.
```
```suggestion
	 * Each device will stream audio to a number of OBS ASIO sources acting as audio clients.
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,140 @@
+	struct asio_data *obs_clients[32];
+	int current_nb_clients;
+
+	/* Output Audio (OBS ==> device).
```
```suggestion
	/* Output Audio (OBS ==> device)
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,140 @@
+	 * for each track. Therefore one has to specify a track index and a channel index for each output channel of
+	 * the device. -1 means no track or a mute channel.
+	 * obs_track[]: array which holds the track index played for each device output channel.
+	 * obs_track_channel[]: array which stores the channel index of an OBS audio track played on an output channel.
```
```suggestion
	 * Each device can be a client to a single OBS output which outputs audio to the said device.
	 * 'output_frames': circular buffer to store the frames which are passed to ASIO devices
	 * Any of the 6 tracks in obs mixer can be streamed to the device. The plugin also allows to select a channel
	 * for each track. Therefore one has to specify a track index and a channel index for each output channel of
	 * the device. -1 means no track or a mute channel.
	 * obs_track[]: array which holds the track index played for each device output channel
	 * obs_track_channel[]: array which stores the channel index of an OBS audio track played on an output channel
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/asio-format.c:
```diff
@@ -0,0 +1,178 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#include "asio-format.h"
+#include "asio.h"
+#include "byteorder.h"
```
```suggestion
******************************************************************************/

#include "asio-format.h"

#include "asio.h"
#include "byteorder.h"
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+struct asio_data {
+	/* common */
+	struct asio_device *asio_device;             // asio device (source plugin: input; output plugin: output)
+	int asio_client_index[MAX_NUM_ASIO_DEVICES]; // index of obs source in device client list
```
```suggestion
	struct asio_device *asio_device;             // ASIO device (source plugin: input; output plugin: output)
	int asio_client_index[MAX_NUM_ASIO_DEVICES]; // index of OBS source in device client list
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+	enum speaker_layout speakers;                // speaker layout
+	int sample_rate;                             // 44100 or 48000 Hz
+	uint8_t in_channels;                         // number of device input channels
+	uint8_t out_channels;                        // output :number of device output channels;
```
```suggestion
	uint8_t out_channels;                        // output: number of device output channels
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+	int sample_rate;                             // 44100 or 48000 Hz
+	uint8_t in_channels;                         // number of device input channels
+	uint8_t out_channels;                        // output :number of device output channels;
+						     // source: number of obs output channels set in OBS Audio Settings
```
The comments here seem odd?

pkviet replied:
the var out_channels is used differently when the struct is for either:

- an ASIO input capture source,
- or an ASIO audio output.
In the first case, out_channels == number of output channels of the device
In the second case, out_channels == number of channels in OBS audio settings

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+	obs_source_t *source;
+	int mix_channels[MAX_AUDIO_CHANNELS]; // stores the channel re-ordering info
+	volatile bool active;                 // tracks whether the device is streaming
+	/* output*/
```
```suggestion
	/* output */
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+	volatile bool active;                 // tracks whether the device is streaming
+	/* output*/
+	obs_output_t *output;
+	uint8_t obs_track_channels;                // number of obs output channels
```
```suggestion
	uint8_t obs_track_channels;                // number of OBS output channels
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+						   // allow for up to 16 channels, should there be a need later to
+						   // expand the channel count (presumbably for broadcast setups).
+						   // Track_index is then stored as 1 << track_index + 4
+						   // so: track 0 = 16, track 1 = 32, etc.
```
Align comments?

pkviet replied:
mmh they're aligned in visual studio for me. clang-format auto changed spaces to tabs so i dunno ?

RytoEX reviewed (changes requested):
Preliminary review. Generally:
* For clarity, prefer enclosing brackets for `if`, `if-else`, `if-else if-else`, `for`, and `while` statements. Secondary preference is to ensure an empty line after an unenclosed statement. If the statement is on its own or is enclosed in a parent statement, it _may_ be left unenclosed, though I still personally prefer just enclosing all of them.
* Line break after a license header before any includes, code, or preprocessor directives such as pragmas.
* Update license header dates to end at 2026.
* Capitalize ASIO and OBS in comments or descriptive text.

pkviet commented:
> Preliminary review. Generally:
> 
> * For clarity, prefer enclosing brackets for `if`, `if-else`, `if-else if-else`, `for`, and `while` statements. Secondary preference is to ensure an empty line after an unenclosed statement. If the statement is on its own or is enclosed in a parent statement, it _may_ be left unenclosed, though I still personally prefer just enclosing all of them.

ok

> * Line break after a license header before any includes, code, or preprocessor directives such as pragmas.

added

> * Update license header dates to end at 2026.

done

> * Capitalize ASIO and OBS in comments or descriptive text.

done

RytoEX commented on frontend/plugins/asio-output-ui/ASIOSettingsDialog.h:
```diff
@@ -0,0 +1,48 @@
+#include <memory>
+#include <QDialog>
+
+#include "ui_output.h"
```
IIRC:
```suggestion
#include "ui_output.h"

#include <obs.hpp>
#include <properties-view.hpp>

#include <QDialog>

#include <memory>
```

pkviet replied:
fixed

RytoEX commented on libobs/obs-source.c:
```diff
@@ -5322,13 +5322,19 @@ static void apply_audio_volume(obs_source_t *source, uint32_t mixers, size_t cha
+	if (vol == 0.0f || mixers == 0 || source->monitoring_type == OBS_MONITORING_TYPE_MONITOR_ONLY) {
 		memset(source->audio_output_buf[0][0], 0,
 		       AUDIO_OUTPUT_FRAMES * sizeof(float) * MAX_AUDIO_CHANNELS * MAX_AUDIO_MIXES);
+		/* We don't silence the extra monitoring mixes so that they can be heard at all times. */
```
Use a C-style comment here:
```suggestion
		// We don't silence the extra monitoring mixes so that they can be heard at all times.
```

pkviet replied:
uh i don't get it; // is c++ not c style

pkviet replied:
it was clarified that it should be /* */ indeed

RytoEX commented on plugins/obs-transitions/transition-stinger.c:
```diff
@@ -137,6 +137,13 @@ static void stinger_update(void *data, obs_data_t *settings)
 	}
 
 	s->monitoring_type = (int)obs_data_get_int(settings, "audio_monitoring");
+	// If the stinger is set to monitor only, don't send audio to output mixers
```
```suggestion
	// If the stinger is set to monitor only, don't send audio to output mixers.
```

pkviet replied:
fixed and also switched to C-style comment

RytoEX commented on plugins/win-asio/data/locale/en-US.ini:
```diff
@@ -0,0 +1,113 @@
+Control.Panel="ASIO Device Control Panel"
+Control.Panel.Hint="If you change any settings of the device, like the buffer, while it's streaming, hit the reload button to reload the driver in case of glitches (due to bad drivers not messaging changes to hosts)."
+Reset.Device="Reload driver"
+Reset.Device.Hint="Reloads manually the driver. This is useful in case of driver settings being modified (sample rate, buffer, etc.)"
```
```suggestion
Reset.Device.Hint="Reloads manually the driver. This is useful in case of driver settings being modified (sample rate, buffer, etc.)."
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-common.h:
```diff
@@ -0,0 +1,47 @@
+/******************************************************************************
+    Copyright (C) 2022-2025 pkv <pkv@obsproject.com>
```
```suggestion
    Copyright (C) 2022-2026 pkv <pkv@obsproject.com>
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device-list.c:
```diff
@@ -0,0 +1,210 @@
+		if (current_asio_devices[i] == NULL)
+			return i;
+	}
+	errorscanner("You have more than 16 asio devices, that's too many !\nShip me some...");
```
```suggestion
	errorscanner("You have more than 16 ASIO devices, that's too many!");
```

This should probably be more descriptive and mention the limit we have in place. The 16 here should probably also be pulled from a defined constant rather than just being a number.

pkviet replied:
switched to: errorscanner("You have more than %i asio devices, that's too many!", MAX_NUM_ASIO_DEVICES);

RytoEX replied:
When I said "more descriptive", I kind of meant specifically indicating that 16 (or whatever the number is defined as) is our maximum limit rather than just "too many".

pkviet replied:
ok, i changed to: You have more than %i asio devices, which is the maximum supported by OBS.

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+/******************************************************************************
+    Copyright (C) 2022-2025 pkv <pkv@obsproject.com>
```
```suggestion
    Copyright (C) 2022-2026 pkv <pkv@obsproject.com>
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,60 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+#pragma once
+#include "asio-common.h"
```
```suggestion
******************************************************************************/

#pragma once

#include "asio-common.h"
```

pkviet replied:
fixed

RytoEX commented on frontend/plugins/asio-output-ui/ASIOSettingsDialog.cpp:
```diff
@@ -0,0 +1,109 @@
+		if (!output_running)
+			output_start();
+	} else if (!wasEmpty && nowEmpty) {
+		// valid device swapped to None: stop if running
```
```suggestion
		// Valid device swapped to None: stop if running
```

pkviet replied:
done

RytoEX commented on frontend/plugins/asio-output-ui/ASIOSettingsDialog.cpp:
```diff
@@ -0,0 +1,109 @@
+		if (output_running)
+			output_stop();
+	} else if (!nowEmpty && newDevice != currentDeviceName) {
+		// output was already started so do nothing ...
```
```suggestion
		// Output was already started so do nothing
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+	list = asio_device_list_create();
+	/*DEBUG: check the plugin when no ASIO drivers are installed
+	 * list->count = 0;
+	 */
```
Don't leave in debug code as comments. Either use a debug define/flag with ifdefs, or remove it.

pkviet replied:
removed

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+		return false;
+	}
+
+	/* we set the name even if the driver fails to load in order to be able to deal with disconnected devices. */
```
```suggestion
	// We set the name even if the driver fails to load in order to be able to deal with disconnected devices.
```

pkviet replied:
kept C-style comment to /* and capitalized as requested

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+	}
+	data->device_name = bstrdup(name);
+
+	/* retrieve the struct asio_device ptr & create the asio_device ptr if NULL */
```
```suggestion
	// Retrieve the struct asio_device ptr and create the asio_device ptr if NULL.
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+	data->asio_device = asio_device_list_attach_device(list, name);
+	struct asio_device *dev = data->asio_device;
+	if (!dev) {
+		errordata("\nFailed to create device % s ", name);
```
```suggestion
		errordata("Failed to create device %s", name);
```

Not sure why there are line breaks before error messages?

pkviet replied:
removed them, when i debugged, i had other log lines hence the line break. Fixed

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+		return false;
+	} else if (!dev->asio) {
+		errordata("\nDriver could not find a connected device or device might already be in use by"
+			  "another host.");
```
You don't have to break strings (clang-format should allow this). It may suggest moving the string to its own line, but I'd prefer the string remain unbroken:
```suggestion
		errordata("Driver could not find a connected device or device might already be in use by another host.");
```

pkviet replied:
done (clang-format kept the string but put it in another line.)

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,830 @@
+	}
+
+	if (!data->is_output) {
+		/* update the device client list if the src was never a client & add src ptr as a client of asio device */
```
```suggestion
		// Update the device client list if the src was never a client and add src ptr as a client of ASIO device.
```

pkviet replied:
fixed

RytoEX reviewed (changes requested):
* Use C-style comments in C.
* Capitalize ASIO in comments and descriptive text.
* Avoid using `&` (ampersands) in descriptive text. Just use "and".

I may need to see some of the logging to understand why there are linebreaks at the beginning of some of them or linebreaks right after a label.

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,209 @@
+#include "ASIOSettingsDialog.h"
+
+#include <obs-frontend-api.h>
+#include <obs-module.h>
+#include <util/util.hpp>
+#include <util/platform.h>
```
```suggestion
#include <obs-frontend-api.h>
#include <obs-module.h>
#include <util/platform.h>
#include <util/util.hpp>
```

pkviet replied:
fixed

RytoEX commented on libobs/obs-source.c:
```diff
@@ -5587,10 +5596,30 @@ void obs_source_set_monitoring_type(obs_source_t *source, enum obs_monitoring_ty
+
+#ifdef _WIN32
+	/* On Windows, assign to the extra ASIO monitoring track (track 7) all sources which have not type
+	 * OBS_MONITORING_TYPE_NONE.*/
```
```suggestion
	 * OBS_MONITORING_TYPE_NONE. */
```

pkviet replied:
applied

RytoEX commented on libobs/obs-source.c:
```diff
@@ -5587,10 +5596,30 @@ void obs_source_set_monitoring_type(obs_source_t *source, enum obs_monitoring_ty
 enum obs_monitoring_type obs_source_get_monitoring_type(const obs_source_t *source)
 {
+#ifdef _WIN32
+	/* If type is OBS_MONITORING_TYPE_NONE, unselect the extra ASIO monitoring track (track 7) on Windows.*/
```
```suggestion
	/* If type is OBS_MONITORING_TYPE_NONE, unselect the extra ASIO monitoring track (track 7) on Windows. */
```

pkviet replied:
applied

RytoEX commented on plugins/win-asio/asio-device.h:
```diff
@@ -0,0 +1,144 @@
+
+#pragma once
+
+#include "asio-common.h"
+#include "asio-format.h"
+#include "asio-compat.h"
+#include "iasiodrv.h"
```
```suggestion
#include "asio-common.h"
#include "asio-compat.h"
#include "asio-format.h"
#include "iasiodrv.h"
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/win-asio.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <obs-frontend-api.h>
+#include <util/platform.h>
+#include <util/threading.h>
+#include <util/darray.h>
+#include <util/deque.h>
```
```suggestion
#include <obs-frontend-api.h>
#include <util/darray.h>
#include <util/deque.h>
#include <util/platform.h>
#include <util/threading.h>
```

pkviet replied:
fixed

RytoEX reviewed (changes requested):
Nits.

pkviet commented:
**Update**
Renamed asio.h → asio-compat.h to avoid confusion with the Steinberg SDK header.

**Clarification**
There is no vendoring of the ASIO SDK: no SDK headers or source files are copied into the tree.
asio-compat.h is a minimal compatibility header that re-declares only the subset of types, enums, and structs required by this implementation (derived from the SDK definitions).
The host interacts directly with the IASIO COM interface via its vtable (pure C implementation).
The SDK’s C++ wrapper layer and additional components are not used, and including the full SDK would introduce a larger, unused API surface without changing how the driver is accessed.

RytoEX commented:
> **Update** Renamed asio.h → asio-compat.h to avoid confusion with the Steinberg SDK header.
> 
> **Clarification** There is no vendoring of the ASIO SDK: no SDK headers or source files are copied into the tree. asio-compat.h is a minimal compatibility header that re-declares only the subset of types, enums, and structs required by this implementation (derived from the SDK definitions). The host interacts directly with the IASIO COM interface via its vtable (pure C implementation). The SDK’s C++ wrapper layer and additional components are not used, and including the full SDK would introduce a larger, unused API surface without changing how the driver is accessed.

I guess my question is: why we can't just include `asio.h` in obs-deps and discard everything else we do not need instead of having a derived version? Are there things in the full header that would cause issues in the implementation in this PR?

pkviet commented:
> > **Update** Renamed asio.h → asio-compat.h to avoid confusion with the Steinberg SDK header.
> > **Clarification** There is no vendoring of the ASIO SDK: no SDK headers or source files are copied into the tree. asio-compat.h is a minimal compatibility header that re-declares only the subset of types, enums, and structs required by this implementation (derived from the SDK definitions). The host interacts directly with the IASIO COM interface via its vtable (pure C implementation). The SDK’s C++ wrapper layer and additional components are not used, and including the full SDK would introduce a larger, unused API surface without changing how the driver is accessed.
> 
> I guess my question is: why we can't just include `asio.h` in obs-deps and discard everything else we do not need instead of having a derived version? Are there things in the full header that would cause issues in the implementation in this PR?

The SDK asio.h can be used, but it requires either patching or adding compile definitions (via CMake) to get correct native types.
I understand the preference for routing third-party material through obs-deps, and I agree with that in general. I did that for vst3 where i do use the sdk extensively (and is typically included in-tree in full by hosts).

But, in this specific case the cost seems disproportionate to what is actually being used. The host only needs a very small subset of ASIO definitions (roughly 130 lines of typedefs / enums / structs derived from the SDK header), and the implementation talks directly to the IASIO COM interface via its vtable. No SDK functions are used.

Moving this through obs-deps would mean just for the sake of a single file:

- adding an ASIO SDK finder
- updating the plugin CMake to consume it
- adding in obs-deps a script  to download the sdk zip from Steinberg (there's no github for the sdk)
- staging a release on my fork and adding a commit in the PR just to expose asio.h header until obs-deps picks it up in a release.

That is a lot of extra infrastructure for a very small compatibility layer  (130 lines out of 1080) that is already local, documented, and limited to the exact subset used by this plugin.
ASIO is widely deployed across hardware, and the subset used here is very stable, so the risk of breakage from maintaining this minimal definition set is extremely low. The SDK has historically evolved via additive changes rather than breaking ABI changes, as backward compatibility is a critical requirement for broad hardware support. I don't see nor foresee any maintenance burden here.

So from a technical point of view, I still think the local asio-compat.h is the more proportionate solution here. It keeps the dependency surface minimal and explicit, without introducing additional build infrastructure or unused API surface.
I don't use the SDK at all, don't call any functions from it, only the COM interface is used.

pkviet commented:
**Update**
Fixed braces due to new Contributing Guidelines.

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,213 @@
+namespace {
+constexpr int MAX_ASIO_DEVICE_CHANNELS = 32;
+constexpr int SHIFT_WINDOW = 100;
+// We use a global context for asio output because we allow a single output device.
```
```suggestion
// We use a global context for ASIO output because we allow a single output device.
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,861 @@
+			data->out_mix_channels[i] = -1;
+		}
+
+		/* allow all tracks for asio output + extra monitoring track */
```
```suggestion
		/* allow all tracks for ASIO output + extra monitoring track */
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+
+const IID IID_IASIO = {0x261a7a60, 0xa003, 0x11d1, {0xa3, 0x90, 0x00, 0x80, 0x5f, 0x08, 0x38, 0x75}};
+
+/* global vars for com initialization in Main thread */
```
```suggestion
/* Global vars for COM initialization in Main thread */
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+
+	if (dev->obs_output_client) {
+		blog(LOG_INFO,
+		     "[asio_output]:\n\tDevice % s removed;\n"
```
Seems like this is a typo?
```suggestion
		     "[asio_output]:\n\tDevice %s removed;\n"
```

pkviet replied:
yes, fixed

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+		Sleep(10);
+
+		if (err == ASE_NoClock && dev->num_clock_sources > 0) {
+			info("trying to set a clock source..");
```
```suggestion
			info("trying to set a clock source...");
```

pkviet replied:
removed the ..

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+	}
+}
+
+/* Next function exists because of shi..y drivers which expect some loading sequence which should be unnecessary in
```
We can probably just say "bad drivers" or "misbehaving drivers" or whatever the actual case may be. Same for elsewhere.

pkviet replied:
ok, opted for bad drivers.

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+}
+
+/* Next function exists because of shi..y drivers which expect some loading sequence which should be unnecessary in
+ * theory. But following JUCE & for extra safety, we do it anyway 'à la Cubase' */
```
Prefer just spelling out "and" over "&".
```suggestion
 * theory. But following JUCE and for extra safety, we do it anyway 'à la Cubase'. */
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+						err = ASIO_Start(dev);
+						Sleep(80);
+						ASIO_Stop(dev);
+						/* We dispose of buffers here and not in the open function bc some
```
```suggestion
						/* We dispose of buffers here and not in the open function because some
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+		os_atomic_set_bool(&dev->need_to_reset, false);
+	}
+
+	/* buffers creation routine; if this fails, try a second time with preferredBufferSize*/
```
```suggestion
	/* buffers creation routine; if this fails, try a second time with preferredBufferSize */
```

pkviet replied:
done

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+	return dev ? dev->obs_output_client : NULL;
+}
+
+/* message calback */
```
```suggestion
/* message callback */
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+		.frames = samps,
+		.timestamp = os_gettime_ns(),
+	};
+	/* pass audio to obs clients & then to the audio pipeline */
```
```suggestion
	/* pass audio to OBS clients & then to the audio pipeline */
```

pkviet replied:
fixed, and changed & to and

RytoEX reviewed (changes requested):
Nits.

RytoEX commented on plugins/win-asio/win-asio.c:
```diff
@@ -0,0 +1,861 @@
+	data->asio_device = asio_device_list_attach_device(list, name);
+	struct asio_device *dev = data->asio_device;
+	if (!dev) {
+		errordata("Failed to create device %s ", name);
```
Nit?

```suggestion
		errordata("Failed to create device %s", name);
```

pkviet replied:
fixed

RytoEX commented on plugins/win-asio/asio-device.c:
```diff
@@ -0,0 +1,1372 @@
+			strncpy(driver_error, "Driver failed to initialize", 30);
+		}
+	}
+	/* that call seems required by sh...y drivers */
```
Just say bad, misbehaving, or non-compliant drivers.

pkviet replied:
replaced with: That call seems required by misbehaving drivers although it is not mandated by the SDK.

RytoEX reviewed (changes requested):
Nits.

RytoEX commented on frontend/plugins/asio-output-ui/asio-ui-main.cpp:
```diff
@@ -0,0 +1,213 @@
+
+namespace {
+constexpr int MAX_ASIO_DEVICE_CHANNELS = 32;
+constexpr int SHIFT_WINDOW = 100;
```
Nit. Per code guidelines, use `kConstantName` for new constants in C++ code.

```suggestion
constexpr int kMaxASIODeviceChannels = 32;
constexpr int kShiftWindow = 100;
```

pkviet replied:
done

RytoEX reviewed (changes requested):
Nit.

pkviet commented:
**Update**
- fixed a series of bugs related to audio output, and swap between devices.
- for audio output, changed the deque to a real circular buffer with atomic accesses.
- the asio audio output relies on obs outputs so it is subject to up to 960 ms of latency; it benefits from the low latency setting in Audio Settings, so I modified the en-US.ini related entry to mention ASIO alongside Decklink.

pkviet commented:
**Update (August 14, 2026)**
- I have made an important architecture change regarding ASIO monitoring. Previously I had added an extra track (so Track 7) to which all monitored sources where automatically streamed. This allowed to have a dedicated track for monitoring. The incentive for that was the fact that the current monitoring API can not be adapted to ASIO because it assumes the audio platform solution (WASAPI, CoreAudio, PulseAudio) will mix all the audio data thrown at it; but ASIO has no such capability. There is however a major drawback which is the audio runs through the main output audio_callback (in obs-audio.c) and therefore can buffer up to 45 audio ticks (960 ms at 48 kHz). And so the main advantage of ASIO with low latency is defeated. Even with the current low latency option (which exists for Decklink outputs for instance), there are at minimum 2 audio ticks so about 40 ms of audio latency.
- I decided to add a new 'monitoring_mix' API to allow the libobs/audio_monitoring to mix monitored sources. Outputs can subscribe to a new raw audio callback 'raw_audio_monitoring' which ticks at 256 samples so about 5 ms.
- I tested against a Yamaha Steinberg IXO 22 sound card. I used a Media Source at 120 fps with an audio burst and a video flash ticking every second and shooted my monitor set to 120 fps with my smartphone camera set at 120 fps. So the resolution for the latency is one video frame so 8.3 ms. I made at least 10 runs of 15 seconds in both WASAPI and ASIO monitoring.
- I found that WASAPI had a latency around 30 - 50 ms, with a latency drifting over time, a fact well-known to OBS users.
- With ASIO, the Yamaha driver was set to a buffer of 128 and a achieved a sub-frame latency (< 8.3 ms) so delay is impossible to spot. I tested also another driver (ESI U22XT at buffer 128) which achieved a latency of about 20 ms against WASAPI (50 -60 ms).

RealIndrit commented on libobs/util/spsc-fifo.h:
```diff
@@ -0,0 +1,218 @@
+	const long writable_frames = fifo->capacity - readable_frames - 1;
+
+	if (frames > writable_frames) {
+		return;
```
I already wrote this through other channels directly , but for transparency and clarity I also put a summary here:

Dropping data like this can be seen as a bad practice for this type of data. (If not intentional for specific purpose), due to this files "Generic" usage placement in the OBS source the assumption is that it could be used outside of the ASIO's feature needs?

I would therefore recommend proper overflow handling for cases outside of expectations today

pkviet replied:
Added a comment on how to use the write function and that it drops data in case of overspill; if one prefers to avoid drops, one should call the spsc_fifo_writable_frames function before calling spsc_fifo_write. If one wants to overwrite data, the deque is better suited but for true spsc we don't want that.
